sklearn/preprocessing/imputation.py

```diff
@@ -115,13 +119,18 @@ class Imputer(BaseEstimator, TransformerMixin):
       contain missing values).
     """
     def __init__(self, missing_values="NaN", strategy="mean",
-                 axis=0, verbose=0, copy=True):
+                 axis=None, verbose=0, copy=True):
         self.missing_values = missing_values
         self.strategy = strategy
-        self.axis = axis
         self.verbose = verbose
         self.copy = copy
 
+        self.axis = axis
+        if axis is not None:
```

Comment:
The warning should be raised in `fit` method. The validation is postponed for the SearhCV: object: http://scikit-learn.org/stable/developers/contributing.html#instantiation